src/analysis/monotone-analyzer-impl.h

```diff
@@ -8,78 +8,83 @@
 
 namespace wasm::analysis {
 template<size_t N>
-BlockState<N>::BlockState(const BasicBlock* underlyingBlock)
+inline BlockState<N>::BlockState(const BasicBlock* underlyingBlock)
   : index(underlyingBlock->getIndex()), cfgBlock(underlyingBlock),
     beginningState(BitsetPowersetLattice<N>::getBottom()),
     endState(BitsetPowersetLattice<N>::getBottom()),
     currState(BitsetPowersetLattice<N>::getBottom()) {}
 
-template<size_t N> void BlockState<N>::addPredecessor(BlockState* pred) {
+template<size_t N> inline void BlockState<N>::addPredecessor(BlockState* pred) {
   predecessors.push_back(pred);
 }
 
-template<size_t N> void BlockState<N>::addSuccessor(BlockState* succ) {
+template<size_t N> inline void BlockState<N>::addSuccessor(BlockState* succ) {
   successors.push_back(succ);
 }
 
-template<size_t N> BitsetPowersetLattice<N>& BlockState<N>::getFirstState() {
+template<size_t N>
+inline BitsetPowersetLattice<N>& BlockState<N>::getFirstState() {
   return beginningState;
 }
 
-template<size_t N> BitsetPowersetLattice<N>& BlockState<N>::getLastState() {
+template<size_t N>
+inline BitsetPowersetLattice<N>& BlockState<N>::getLastState() {
   return endState;
 }
 
 // In our current limited implementation, we just update a new live variable
 // when it it is used in a get or set.
-template<size_t N> void BlockState<N>::visitLocalSet(LocalSet* curr) {
-  currState.value[curr->index] = true;
+template<size_t N> inline void BlockState<N>::visitLocalSet(LocalSet* curr) {
+  currState.value[curr->index] = false;
 }
 
-template<size_t N> void BlockState<N>::visitLocalGet(LocalGet* curr) {
+template<size_t N> inline void BlockState<N>::visitLocalGet(LocalGet* curr) {
   currState.value[curr->index] = true;
 }
 
-template<size_t N> void BlockState<N>::transfer() {
+template<size_t N> inline void BlockState<N>::transfer() {
   if (cfgBlock->size() == 0) {
     beginningState.getLeastUpperBound(endState);
     return;
   }
 
   // compute transfer function for all expressions in the CFG block
   auto cfgIter = cfgBlock->rbegin();
-  currState.copy(endState);
+  currState = endState;
 
   while (cfgIter != cfgBlock->rend()) {
     // run transfer function.
     BlockState<N>::visit(*cfgIter);
     ++cfgIter;
   }
-  beginningState.copy(currState);
+  beginningState = currState;
```

Comment:
I'm surprised that we're setting `beginningState` to the result of applying the transfer function. Shouldn't we leave this block's `beginningState` alone and propagate the results to the successor blocks instead?

Reply:
Ah, I see that the naming here assumes a backward analysis. LGTM for now, although we'll want to fix that later.